### PR TITLE
First implementation #1

### PR DESCRIPTION
**Pull request by abbadon1334**
Working on #1

### CODE REVIEW

File: src/Mailer/AbstractMailer.php [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+                $mail,
+                'cc',
+                function ($address): void {
+                    $this->phpmailer->addCC($address->get('email'), $address->get('name'));
+                }
+            );
+
+            $this->addAddress(
+                $mail,
+                'replyto',
```
Comment (DarkSide666) [inline src/Mailer/AbstractMailer.php]: shouldn't there be `bcc` ?

File: src/Mailer/Gmail.php [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+    public function __construct(array $defaults = [])
+    {
+        if (empty($defaults['password']) || empty($defaults['username'])) {
+            throw new Exception('username and password must be defined in injection array');
+        }
+
+        parent::__construct($defaults);
+
+        $this->phpmailer->isSMTP();
```
Comment (DarkSide666) [inline src/Mailer/Gmail.php]: no needs for this line because it's already in Mailer/SMTP class

File: src/Model/Mail.php [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        return $this;
+    }
+
+    public function send(?Outbox $outbox = null): void
+    {
+        if (null === $outbox && null !== $this->app && method_exists($this->app, 'getOutbox')) {
+            $outbox = $this->app->getOutbox();
+        }
+
+        $outbox->send($this);
```
Comment (DarkSide666) [inline src/Model/Mail.php:185]: here $outbox still can be `null` so need some additional checking i think

File: src/Outbox.php [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    }
+    use DIContainerTrait;
+    use FactoryTrait;
+
+    /**
+     * Mailer.
+     *
+     * @var MailerInterface
+     */
+    private $mailer;
```
Comment (DarkSide666) [inline src/Outbox.php]: please use `protected` instead of `private` properties so this class can be extended more easily

File: src/Outbox.php [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    /**
+     * @throws Exception
+     */
+    public function init(): void
+    {
+        $this->_init();
+
+        if (null !== $this->app) {
+            $this->app->addMethod(
+                'getOutBox',
```
Comment (DarkSide666) [inline src/Outbox.php]: Should be `getOutbox` not `getOutBox` (letter B case)

**Review by DarkSide666 — CHANGES_REQUESTED** [review]
Found some bugs, but overall looks nice.

File: src/Outbox.php [hunk 6]
```diff
@@ -127,9 +122,15 @@ public function send(Mail $mail): Mail
         $mail->hook('afterSend');
     }
 
-    public function callableSend(callable $send): void
+    /**
+     * @throws Exception
+     */
+    private function validateOutbox(): void
```
Comment (DarkSide666) [inline src/Outbox.php]: pls `protected`

**Review by DarkSide666 — APPROVED** [review]
nice

File: src/Outbox.php (hunk 6 above)
Comment (abbadon1334) [reply src/Outbox.php]: done

**Review by romaninsh — APPROVED** [review]
Thanks so much for contributing this!

**Pull request merged**